webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/policies/DeploymentService.scala

```diff
@@ -897,102 +943,149 @@ object BuildNodeConfiguration extends Loggable {
+
+                                /*
+                                 * Clearly, here, we are evaluating parameters, and we are not using that just after in the
+                                 * variable expansion, which mean that we are doing the same work again and again and again.
+                                 * Moreover, we also are evaluating again and again parameters whose context ONLY depends
+                                 * on other parameter, and not node config at all. Bad bad bad bad.
+                                 * TODO: two stages parameter evaluation
+                                 *  - global
+                                 *  - by node
+                                 *  + use them in variable expansion (the variable expansion should have a fully evaluated InterpolationContext)
+                                 */
+                                parameters     <- context.parameters.accumulate { case (name, param) =>
+                                                    for {
+                                                      p <- param(context)
+                                                    } yield {
+                                                      (name, p)
+                                                    }
+                                                  }.mapError(_.deduplicate)
+
+                    ////////////////////////////////////////////  fboundedDraft ////////////////////////////////////////////
+
+                                t1_2           <- nanoTime
+                                _              <- counters.sumTimeParameter.update(_ + t1_2 - t1_1)
+                                boundedDrafts  <- filteredDrafts.accumulate { draft =>
+                                                    (for {
+                                                      _          <- counters.nbBoundDraft.update(_ + 1)
+                                                      ret        <- (for {
+                                                                      //bind variables with interpolated context
+                                                                      t2_0              <- nanoTime
+                                                                      expandedVariables <- draft.variables(context)
+                                                                      t2_1              <- nanoTime
+                                                                      _                 <- counters.sumTimeExpandeVar.update(_ + t2_1 - t2_0)
+                                                                      // And now, for each variable, eval - if needed - the result
+                                                                      expandedVars      <- expandedVariables.accumulate { case (k, v) =>
+                                                                                             //js lib is specific to the node os, bind here to not leak eval between vars
```

Comment:
i think this could be done outside the flteredDraft.accumulate, as we are dealing node per node
but i don't think it would change much perf, but surely allocate less objects